src/pages/sign-in-page/sign-in-page.tsx

```diff
@@ -2,42 +2,94 @@ import React, { useState } from 'react';
 import { useNavigate } from 'react-router-dom';
 import { RegistrationPanel } from '../../components/registration-panel/registration-panel';
 import { TextBox } from '../../components/text-box/text-box';
+import { isEmptyObject } from '../../helpers/validator';
 import { useUserClient } from '../../hooks/api/use-user-client';
 import { paths } from '../../routing/paths';
+import './sign-in-page.scss';
+
+interface SignInFormError {
+  email?: string;
+  password?: string;
+  generic?: string;
+}
 
 export const SignInPage = () => {
   const userClient = useUserClient();
   const navigate = useNavigate();
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
 
+  const [formError, setFormError] = useState<SignInFormError>();
   const [isSubmitting, setIsSubmitting] = useState(false);
 
   return (
     <RegistrationPanel
+      className="sign-in-panel"
       formHeader="sign in to your account"
       submitLabel="sign in"
       swapPanelLabel="create an account"
       submitLabelLoading={isSubmitting}
       onSubmitClick={handleSubmitClick}
       onSwapPanelClick={handleSwapPanelClick}
     >
-      <TextBox label="email" variant="dark-white" value={email} onChange={setEmail} />
-      <TextBox
-        label="password"
-        variant="dark-white"
-        inputType="password"
-        value={password}
-        onChange={setPassword}
-      />
+      <div className="field-container centered">
+        {getLabelError('', (formError) => formError.generic)}
+      </div>
+      <div className="field-container">
+        <TextBox
+          label="email"
+          autoComplete="email"
+          variant="dark-white"
+          value={email}
+          onChange={setEmail}
+        />
+        {getLabelError('', (formError) => formError.email)}
+      </div>
+      <div className="field-container">
+        <TextBox
+          label="password"
+          autoComplete="current-password"
+          variant="dark-white"
+          inputType="password"
+          value={password}
+          onChange={setPassword}
+        />
+        {getLabelError('', (formError) => formError.password)}
+      </div>
     </RegistrationPanel>
   );
 
+  function handleSwapPanelClick() {
+    navigate(paths.signUp);
+  }
+
   async function handleSubmitClick() {
     if (isSubmitting) {
       return;
     }
 
+    setFormError(undefined);
+
+    // simple client side verifications
+    {
+      const newFormError: SignInFormError = {};
+      console.log(email.length, password.length);
+      if (email.length === 0) {
+        newFormError.email ??= 'Email should not be empty.';
+      }
+      if (!email.includes('@')) {
+        newFormError.email ??= `This doesn't appear to be a valid email.`;
```

Comment:
Didn't know `??=` was a think either, lol

Reply:
yeah!! it saves a lot of headache